src/app/icd/client/DefaultICDClientStorage.cpp

```diff
@@ -0,0 +1,397 @@
+/*
+ *    Copyright (c) 2023 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include "DefaultICDClientStorage.h"
+#include <iterator>
+#include <lib/core/Global.h>
+#include <lib/support/Base64.h>
+#include <lib/support/CodeUtils.h>
+#include <lib/support/DefaultStorageKeyAllocator.h>
+#include <lib/support/SafeInt.h>
+#include <lib/support/logging/CHIPLogging.h>
+
+#define FABRIC_INDEX_TLV_SIZE 2
+
+namespace chip {
+namespace app {
+CHIP_ERROR DefaultICDClientStorage::UpdateFabricList(FabricIndex fabricIndex)
+{
+    for (auto & fabric : mFabricList)
+    {
+        if (fabric == fabricIndex)
+        {
+            return CHIP_ERROR_INVALID_ARGUMENT;
+        }
+    }
+
+    mFabricList.push_back(fabricIndex);
+
+    Platform::ScopedMemoryBuffer<uint8_t> backingBuffer;
+    size_t counter      = mFabricList.size() + 1;
+    size_t total = FABRIC_INDEX_TLV_SIZE * counter;
```

Comment:
This doesn't account for structure overhead. Please do not use the +1 above to do so.

Reply:
agree, updated, thanks